### ISSUE TITLE
Training

### ISSUE THREAD

**Issue by ymliu-felix** [TinyLLaVA/TinyLLaVA_Factory#211]
Why does training work normally when I use OpenELM-270M, but when I use version 3B, the loss is always 0?

**Comment by BobYue-01** [comment]
Thanks for the report.

A few things would help us narrow this down:

- Which TinyLLaVA commit/version are you using?
- What exact model structure are you training? (A pure language model, or a LLaVA-like model assembled through the current registry system?)
- Could you provide a minimal reproduction (training command, config, and dataset format)?
- Are there any warnings during preprocessing or tokenization?

One thing to note is that the current `main` branch does not yet provide standardized support for pure text LLMs. We are actively addressing this in the ongoing refactor on the `develop` branch by aligning more closely with standard Hugging Face interfaces and contracts. As a result, some assumptions in your current training pipeline may work with OpenELM-270M but not necessarily with larger OpenELM variants such as the 3B model.

That said, this report is still valuable. Even if the root cause turns out to be related to limitations in the current implementation, identifying where and why the behavior diverges can help validate and improve the ongoing refactor.

If you can provide a minimal reproducible example, it will help us determine whether this is:

- a label masking issue (e.g. all labels becoming `-100`),
- a template or tokenization issue,
- a model loading or configuration mismatch,
- or another incompatibility in the current training pipeline.

Any logs, configs, or reproduction steps would be greatly appreciated.

**Comment by ymliu-felix** [comment]
1. I just used the tinyllava in the Main brach. 
2. I just change the model from apple/OpenELM-270M-Instruct to apple/OpenELM-3B-Instruct in the provided train_openelm.sh to train on the LLaVA dataset.
3. There is no warnings.
4. When I use apple/OpenELM-11_B-Instruct, the loss is always 0 for several (about 3) steps, and then it will be normal with the initial gradient of about 84. So I guess it is because of exploding gradients? When I use the version of 270M, it is normal without any problem.

Here is my script:

```bash
DATA_PATH=blip_laion_cc_sbu_558k.json
FINETUNE_DATA_PATH=llava_v1_5_mix665k.json
IMAGE_PATH=images
FINETUNE_IMAGE_PATH=general_vqa_train

LLM_VERSION=apple/OpenELM-1_1B-Instruct
VT_VERSION=google/siglip-so400m-patch14-384
VT_VERSION2=""
CN_VERSION=mlp2x_gelu
CONV_VERSION=llama
VERSION=elm_base
TRAIN_RECIPE=common
MODEL_MAX_LENGTH=2048


bash scripts/train/openelm/pretrain_openelm.sh "$DATA_PATH" "$IMAGE_PATH" "$LLM_VERSION" "$VT_VERSION" "$VT_VERSION2" "$CN_VERSION" "$VERSION" "$TRAIN_RECIPE" "$MODEL_MAX_LENGTH"
```

```bash
#!/bin/bash

if [ $# -ne 9 ]; then
    echo "Usage: $0 <DATA_PATH> <IMAGE_PATH> <LLM_VERSION> <VT_VERSION> <VT_VERSION2> <CN_VERSION> <VERSION> <TRAIN_RECIPE> <MODEL_MAX_LENGTH>"
    exit 1
fi

# Assign the arguments to variables
DATA_PATH="$1"
IMAGE_PATH="$2"
LLM_VERSION="$3"
VT_VERSION="$4"
VT_VERSION2="$5"
CN_VERSION="$6"
VERSION="$7"
TRAIN_RECIPE="$8"
MODEL_MAX_LENGTH="$9"

VT_VARIANT="${VT_VERSION#*/}"
LLM_VARIANT="${LLM_VERSION#*/}"

deepspeed --include localhost:0,1,2,3 --master_port 29503 tinyllava/train/train.py \
    --deepspeed ./scripts/zero2.json \
    --data_path  $DATA_PATH\
    --image_folder $IMAGE_PATH \
    --is_multimodal True \
    --conv_version pretrain \
    --model_name_or_path $LLM_VERSION \
    --vision_tower $VT_VERSION \
    --vision_tower2 "$VT_VERSION2" \
    --connector_type $CN_VERSION \
    --mm_vision_select_layer -2 \
    --mm_vision_select_feature cls_patch \
    --image_aspect_ratio square \
    --fp16 True \
    --tokenizer_name_or_path models/Llama-2-7b-hf \
    --training_recipe $TRAIN_RECIPE \
    --tune_type_llm frozen \
    --tune_type_vision_tower frozen \
    --tune_vision_tower_from_layer 0 \
    --tune_type_connector full \
    --output_dir tiny-llava-${LLM_VARIANT}-${VT_VARIANT}-${VERSION}-pretrain \
    --num_train_epochs 1 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 8 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 50000 \
    --save_total_limit 1 \
    --learning_rate 1e-3 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 False \
    --model_max_length $MODEL_MAX_LENGTH \
    --gradient_checkpointing True \
    --dataloader_num_workers 8 \
    --lazy_preprocess True \
    --report_to tensorboard \
    --tokenizer_use_fast False \
    --run_name tiny-llava-${LLM_VARIANT}-${VT_VARIANT}-${VERSION}-pretrain
```

**Comment by BobYue-01** [comment]
Thanks for the detailed reproduction steps.

I was able to reproduce the issue with OpenELM-3B using the provided pretraining setup. Under the same configuration, OpenELM-270M and OpenELM-1.1B appear to train normally, while OpenELM-3B consistently reports from the beginning of training:

```text
loss: 0.0
grad_norm: 0.0
```

A few observations so far:

- The issue is reproducible on our side.
- It does not appear to depend on the number of GPUs used.
- The behavior is the same with both ZeRO-2 and ZeRO-3.
- The `torch.utils.checkpoint` warning about `None of the inputs have requires_grad=True` also appears with OpenELM-270M, which trains normally, so it is likely not the direct cause of the issue.

At this point, the problem seems specific to the interaction between OpenELM-3B and the current TinyLLaVA training pipeline rather than a general DeepSpeed configuration issue.

We'll continue investigating. The next steps are to compare the OpenELM-3B path against the working 270M/1.1B configurations and verify whether supervision signals and gradients are being propagated correctly throughout the training graph.

Thanks again for the report — having a reproducible case is very helpful for tracking this down.

**Comment by BobYue-01** [comment]
Thanks for the detailed report. We were able to locate the issue.

### Workaround

For `apple/OpenELM-3B-Instruct`, please use BF16 instead of FP16:

```diff
-     --fp16 True
+     --bf16 True
```

On our side, switching from FP16 to BF16 immediately resolves the issue and produces normal finite losses and gradients.

### What we found

After isolating the OpenELM language-model forward path from the full TinyLLaVA training pipeline, we found that the problem is not caused by label masking, datasets, multimodal preprocessing, connector training, or DeepSpeed topology.

Minimal observations:

* `apple/OpenELM-270M-Instruct` + FP16: works
* `apple/OpenELM-1_1B-Instruct` + FP16: works
* `apple/OpenELM-3B-Instruct` + FP16: NaN logits, NaN loss, NaN gradients
* `apple/OpenELM-3B-Instruct` + BF16: finite logits, finite loss, finite gradients

The issue can be reproduced using a standalone `OpenELMForCausalLM` forward/backward pass without any TinyLLaVA-specific components.

<details>

<summary>Minimal code</summary>

```python
# minimal_openelm_3b_dtype_check.py

import os
import torch
from tinyllava.model.llm.openelm import OpenELMForCausalLM

torch.manual_seed(42)

hf_home = os.environ.get("HF_HOME", None)
snapshot_path = "hub/models--apple--OpenELM-3b-Instruct/snapshots/8288250ae190b81ed7ad515ea661e880b9cde2ba"
model_path = os.path.join(hf_home, snapshot_path) if hf_home is not None else snapshot_path

for dtype in (torch.float16, torch.bfloat16):
    torch.cuda.empty_cache()
    print(f"\n== {dtype} ==")

    model = OpenELMForCausalLM.from_pretrained(
        model_path,
        torch_dtype=dtype,
    ).cuda().train()
    model.config.use_cache = False

    ids = torch.randint(10, 1000, (1, 64), device="cuda")
    inputs = model.get_input_embeddings()(ids).detach().clone().requires_grad_(True)
    labels = ids.clone()
    labels[:, :8] = -100

    out = model(inputs_embeds=inputs, labels=labels, use_cache=False)
    print("loss:", out.loss)
    print("loss finite:", torch.isfinite(out.loss).item())
    print("logits abs max:", out.logits.detach().abs().max())

    out.loss.backward()
    print("grad finite:", torch.isfinite(inputs.grad).all().item())
    print("grad abs sum:", inputs.grad.detach().abs().sum())
```

Results:

```text
== torch.float16 ==
loss: tensor(nan)
loss finite: False
logits abs max: tensor(nan)
grad finite: False
grad abs sum: tensor(nan)

== torch.bfloat16 ==
loss: tensor(18.3750)
loss finite: True
logits abs max: tensor(49.7500)
grad finite: True
grad abs sum: tensor(1760.)
```

</details>

### Interpretation

The current evidence points to a numerical stability issue specific to OpenELM-3B running in FP16.

We checked the official OpenELM configuration and noticed that the released OpenELM family checkpoint declares in [its model config](https://huggingface.co/apple/OpenELM-3B-Instruct/blob/main/config.json):

```json
"torch_dtype": "bfloat16"
```

This is consistent with our findings.

Given that OpenELM uses RMSNorm, SwiGLU (`ffn_with_glu=true`), and a BF16-native checkpoint configuration, BF16 should be considered the intended training dtype for this model family. Smaller variants (270M and 1.1B) happen to remain stable under FP16 in our early tests, but the 3B model exceeds the numerical range that FP16 can safely represent.

In other words, the observed `loss=0` behavior is not the root cause itself. The underlying issue is FP16 numerical overflow, which DeepSpeed subsequently handles by skipping invalid updates.

### Follow-up

The current `main` branch was originally developed primarily as a lightweight research reproduction codebase for the TinyLLaVA paper, and several training scripts use FP16 as a default setting. This appears to be unsafe for OpenELM-3B.

We will update the scripts accordingly and add a clearer note/check. Longer term, the ongoing `develop` branch refactor is moving toward a more standardized Hugging Face-style configuration system such as `dtype="auto"` where dtype selection becomes model-aware rather than being hardcoded in shell scripts.

Thanks again for reporting this and providing the reproduction details. The report helped identify a real compatibility issue in the current training scripts.
